fastfield_codecs/src/null_index/sparse.rs

```diff
@@ -0,0 +1,534 @@
+        u16::from_le_bytes(bytes)
+    }
+
+    #[inline]
+    #[allow(clippy::comparison_chain)]
+    // Looks for the element in the block. Returns the positions if found.
+    fn binary_search(&self, data: &[u8], target: u16) -> Option<u16> {
+        let mut size = self.num_vals;
+        let mut left = 0;
+        let mut right = size;
+        while left < right {
+            let mid = left + size / 2;
+
+            let mid_val = self.value_at_idx(data, mid);
+
+            if target > mid_val {
+                left = mid + 1;
+            } else if target < mid_val {
+                right = mid;
+            } else {
+                return Some(mid);
+            }
+
+            size = right - left;
+        }
+        None
+    }
+}
+
+const SERIALIZED_BLOCK_METADATA_SIZE: usize = 4;
+
+fn deserialize_sparse_codec_block(data: &[u8]) -> Vec<SparseCodecBlock> {
+    // The number of vals so far
+    let mut offset = 0;
+    let mut sparse_codec_blocks = Vec::new();
+    let num_blocks = u16::from_le_bytes([data[data.len() - 2], data[data.len() - 1]]);
+    let block_data_index_start =
+        data.len() - 2 - num_blocks as usize * SERIALIZED_BLOCK_METADATA_SIZE;
+    for block_num in 0..num_blocks as usize {
+        let block_data_index = block_data_index_start + SERIALIZED_BLOCK_METADATA_SIZE * block_num;
+        let block_idx = u16::from_le_bytes([data[block_data_index], data[block_data_index + 1]]);
+        let num_vals = u16::from_le_bytes([data[block_data_index + 2], data[block_data_index + 3]]);
+        while block_idx != sparse_codec_blocks.len() as u16 {
+            // fill up empty blocks
+            sparse_codec_blocks.push(SparseCodecBlock::empty_block(offset));
+        }
+        let block = SparseCodecBlock { num_vals, offset };
+        sparse_codec_blocks.push(block);
+        offset += num_vals as u32;
+    }
+    sparse_codec_blocks.push(SparseCodecBlock::empty_block(offset));
+    sparse_codec_blocks
+}
+
+const ELEMENTS_PER_BLOCK: u32 = u16::MAX as u32;
+
+/// Splits a idx into block index and value in the block
+fn split_in_block_idx_and_val_in_block(idx: u32) -> (u16, u16) {
+    let val_in_block = (idx % u16::MAX as u32) as u16;
+    let block_idx = (idx / u16::MAX as u32) as u16;
+    (block_idx, val_in_block)
+}
+
+impl SparseCodec {
+    /// Open the SparseCodec from OwnedBytes
+    pub fn open(data: OwnedBytes) -> Self {
+        let blocks = deserialize_sparse_codec_block(&data);
+        Self { data, blocks }
+    }
+
+    #[inline]
+    /// Check if value at position is not null.
+    pub fn exists(&self, idx: u32) -> bool {
+        let (block_idx, val_in_block) = split_in_block_idx_and_val_in_block(idx);
+        // There may be trailing nulls without data, those are not stored as blocks. It would be
+        // possible to create empty blocks, but for that we would need to serialize the number of
+        // values or pass them when opening
```

Comment:
I'd rather do that actually.

Reply:
It would add some fixed bytes to every sparse codec field